BASS-356/visual updates

jomurgel opened the pull request:
- Updates webpack to utilize the `DependencyExtractionWebpackPlugin` to import dependencies from packages rather than global wp object.
- Utilizes core Gutenberg elements to style more inline with editor.

![CleanShot 2022-02-18 at 11 37 07](https://user-images.githubusercontent.com/5230729/154742636-6c17e45f-a271-482e-9c28-009d99e51eaa.png)

NOTE: Large diff is the result of package(lock|json).

kevinfodness commented on client/src/components/BylineMetaBox/BylineProfiles.js:
```diff
@@ -131,92 +137,102 @@ class BylineProfiles extends Component {
+              )}
+              renderItem={(item, isHighlighted) => (
+                <div
+                  className={`item ${isHighlighted ?
+                    'item-highlighted' : ''}`}
```
This is a good use case for `classnames` and conditional classes (e.g., `classnames('item', { 'item-highlighted': isHighlighted })`

kevinfodness commented on client/src/components/BylineMetaBox/BylineProfiles.js:
```diff
@@ -131,92 +137,102 @@ class BylineProfiles extends Component {
+                }));
+              }}
+              onChange={(event, value) => {
+                clearTimeout(this.delay);
+                this.setState({
+                  search: value,
+                });
+
+                this.delay = setTimeout(() => {
+                  this.doProfileSearch(value);
+                }, 500);
```
Possible to use a hook like `useDebounce` here instead?

jomurgel replied:
Perhaps, I'll look into it.

kevinfodness commented on client/src/components/BylineMetaBox/BylineProfiles.js:
```diff
@@ -131,92 +137,102 @@ class BylineProfiles extends Component {
+              className="components-base-control__label"
+              htmlFor="byline_freeform"
+            >
+              {window.bylineData.addFreeformlabel}
```
Shouldn't access the `window` object here - if there is configuration data that gets passed by WP then that should either flow down into this component as a prop or it should be extracted to a function that is responsible for getting a configuration value and interacting with `window`.

jomurgel replied:
Looks like we're using localize script here which is feeding the slotfill/components a bunch of data. Will have to think about how best to restructure this. I might open a new branch to wrangle this and a few other improvements and figure out how best to handle this...

kevinfodness commented on client/src/components/BylineMetaBox/BylineProfiles.js:
```diff
@@ -131,92 +137,102 @@ class BylineProfiles extends Component {
+                onChange={(e) => {
+                  this.setState({ value: e.target.value });
+                }}
+                placeholder={window.bylineData.addFreeformPlaceholder}
```
Same comment here and throughout.

kevinfodness commented on client/src/components/BylineSlot/bylineFreeform/index.jsx:
```diff
@@ -21,30 +22,43 @@ const BylineFreeform = (props) => {
+      className="freeform-controls components-base-control__field"
+      style={{ marginTop: 15 }}
+    >
+      {/* eslint-disable jsx-a11y/label-has-for */}
```
The label is associated with an input with `for` and `id` so I'm not clear why this exclusion needs to be present - it can likely be removed. Check the other components in this PR as well.

kevinfodness reviewed (approved):
A few suggestions, no blockers 🍣

jomurgel commented:
Made several updates per suggestions above as well as refactored a few class components in service of improving some things. Also renamed elements from camel to kabab case and moved some util functions until the respective folder.

Additionally made a few tweaks in service of the classic editor which is now space appropriately. 
![CleanShot 2022-02-22 at 14 51 59](https://user-images.githubusercontent.com/5230729/155235448-a1523091-53b8-4694-8bb8-89cb407b1773.png)

NOTE: The giant change in diff is a result of the package lock and file renames.

dlh01 commented on client/src/components/byline-autocomplete/index.jsx:
```diff
@@ -0,0 +1,121 @@
+
+  const doProfileSearch = (fragment) => {
+    fetch(
+      `${profilesApiUrl}?s=${fragment}`,
```
Given that `@wordpress` scripts will be available, any reason to not use `@wordpress/url.addQueryArgs()` here so that `?` doesn't have to be manually concatenated and so that any existing query parameters in `profilesApiUrl` are accommodated?

jomurgel replied:
I guess I'm not sure if this is something we'd ever have to worry about. Since we're hitting these API URLs directly from the root. Though I get the idea. I might make a note of that in an issue for improvement for now since we're waiting on a few of these updates.

jomurgel replied:
https://github.com/alleyinteractive/byline-manager/issues/62

dlh01 commented on client/src/components/byline-profiles.js/index.js:
```diff
@@ -0,0 +1,263 @@
+      name={`byline_entry[${count}][value]`}
+      value={bylineId || name}
+    />
+    { image && <img src={image} alt={name} /> }
```
```suggestion
    { image ? <img src={image} alt={name} /> : null }
```
seems to be the fashion these days.

jomurgel replied:
Yup, based on this article: https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

dlh01 commented on client/src/components/byline-profiles.js/index.js:
```diff
@@ -0,0 +1,263 @@
+  profiles: PropTypes.arrayOf(PropTypes.shape(BYLINE_PROFILE_SHAPE)).isRequired,
+};
+
+export default BylineProfiles;
```
Any particular reason this file is in a directory ending with `.js`?

dlh01 commented on client/src/components/byline-profiles.js/index.js:
```diff
@@ -0,0 +1,263 @@
+  const doProfileSearch = (fragment) => {
+    fetch(
+      `${profilesApiUrl}?s=${fragment}`,
+    )
```
Same suggestion as above.

dlh01 commented on client/src/components/user-link-metabox/index.js:
```diff
@@ -0,0 +1,138 @@
+
+  const doUserSearch = (fragment) => {
+    fetch(
+      `${usersApiUrl}?s=${fragment}&post=${postId}`,
```
Same as above.

dlh01 commented on client/src/components/user-link-metabox/index.js:
```diff
@@ -0,0 +1,138 @@
+  }, [debouncedSearchString]);
+
+  useEffect(() => {
+    setUser(rawUser);
```
Could `rawUser` be passed to `useState()`?

jomurgel replied:
You almost never want to pass a value directly to state unless you know for sure that it's available on mount. In this case, we could return a null or empty value on mount so should set it in a useEffect hook instead.

dlh01 commented on client/src/components/user-link-metabox/index.js:
```diff
@@ -0,0 +1,138 @@
+      {user && user.id && user.name ? (
+        <p className="current-user-link">
+          {`${linkedToLabel} `}
+          <a href={`/wp-admin/user-edit.php?user_id=${user.id}`}>
```
I don't think this path will work as expected in multisite installations where sites use subdirectories, a not-uncommon setup. It would link me from `mainsite.com/subsite/wp-admin/...` to `mainsite.com/wp-admin/user-edit.php`, which I might not have permission to access.

I can see that this issue existed in the old component, so I suppose it doesn't need to be a blocker now, but if you're so inclined, it might be fixable by including the URL to the user edit screen in the response from the users API endpoint (unless that endpoint is publicly accessible, in which case, let's not add an admin link to it) or by exporting the base admin URL from the server to the browser.

jomurgel replied:
👍 I think this might also be a good improvement. For all intent and purpose, nothing has changed from what we had before other than these components have been renamed and refactored (in some cases).

dlh01 commented on inc/assets.php:
```diff
@@ -106,7 +106,7 @@ function localize_admin_script( $handle ) {
 		'bylineData',
 		[
-			'addAuthorLabel'         => __( 'Search for an author to add to the byline', 'byline-manager' ),
+			'addAuthorLabel'         => __( 'Search for a byline author', 'byline-manager' ),
```
I'd be curious to know whether this change came out of a specific experience with users of the plugin?

To me, from the outside, the string previously expressed two common-ish concepts ("an author" and "the byline") but now expresses one new-ish concept ("a byline author"). Bylines might also have multiple components of which authors are only one (see #8), which is encompassed a little more cleanly with the current string, in my opinion. So, IMO, some precision is being lost here. 

(Same applies to the "Author Bylines" string earlier in the PR.)

jomurgel replied:
You're right that is confusing. The confusing part for clients was not really understanding the difference between the author and the byline. The author people a person and the byline being the container for housing authors. This has been updated to be "Search for an author" which is essentially what's happening here.

Part of the update was to reduce the line length too.

dlh01 reviewed (approved):
🌳 This looks like a significant upgrade to the plugin!
